### DISCUSSION TITLE
Allow pixel-specific processors

### DISCUSSION THREAD

**Discussion by Sergio0694** [SixLabors/ImageSharp#1491] [Ideas] [answered]
> **NOTE:** follow up from Discord, pinging @JimBobSquarePants, @antonfirsov, @john-h-k, @DaZombieKiller.

### Overview

Opening this discussion to propose allowing 3rd party to properly implement pixel-specific pixel processors.
This would primarily involve the following changes:
- Exposing `IImageProcessingContext<TPixel>`
- Expose `Clone` and `Mutate` extensions taking `Action<IImageProcessingContext<TPixel>>`
- Expose `Clone` and `Mutate` extensions taking `params IImageProcessor<TPixel>[]`

The reason for this is that right now consumers of the library wanting to implement custom image processors are pretty much forced into necessarily writing pixel-agnostic ones, which is not always ideal, or doable at all without big compromises. This is particularly an issue in cases such as when doing interop with the GPU (eg. like I'm doing in ComputeSharp), because not all pixel formats are supported by the GPU.

Right now this means even when implementing a GPU-based image processor, I have to still do the pixel conversion to `Vector4` on the CPU, which is much slower and completely prevents me from just passing the image on the GPU and using the implicit mapping (aka `Texture2D<unorm float4>` in HLSL), which would mean the GPU would take care of doing conversions for me. This also means I end up having to allocate much bigger buffers on the GPU (as in, `Vector4` ones instead of just packed `uint`).

Pixel-format agnostic processors are absolutely great, but I do think they're not necessarily the best approach in all situations, and the current APIs are effectively causing consumers to lose type information when they're actually working with pixel-specific images. This is not ideal and can be a problem in scenarios like the one mentioned above. I feel like it'd actually make ImageSharp more flexible in general and usable in scenarios where it's not really ideal today.

### Practical example

Consider this example of doing image processing with ComputeSharp, from an image loaded with System.Drawing:

```csharp
Bitmap bitmap = new("cat.jpg");
BitmapData data = bitmap.LockBits(
    new Rectangle(default, bitmap.Size),
    ImageLockMode.ReadWrite,
    PixelFormat.Format32bppArgb);
Span<Bgra32> span = new(data.Scan0.ToPointer(), bitmap.Width * bitmap.Height);

try
{
    using ReadWriteTexture2D<Bgra32, Vector4> texture =
        Gpu.Default.AllocateReadWriteTexture2D<Bgra32, Vector4>(span, bitmap.Width, bitmap.Height);

    // Process the image on the GPU...

    texture.GetData(span);
}
finally
{
    bitmap.UnlockBits(data);
}

bitmap.Save("cat_processed.jpg");
```

You can see here that since I know for a fact the image will be loaded as `BGRA32`, I can just directly copy that (packed) data to GPU memory, and then have my shader work on it directly mapping those values to `Vector4` values. The GPU will automatically perform the pixel conversion whenever the shader reads/writes texture elements. The CPU is doing no work at all.

This is not really doable with ImageSharp, unfortunately.
Imagine I wanted to write a custom processor to work the same way, like this:

```csharp
public sealed class ComputeSharpProcessor : ImageProcessor<Rgba32>
{
    public ComputeSharpProcessor(Configuration configuration, Image<Rgba32> source, Rectangle sourceRectangle)
        : base(configuration, source, sourceRectangle)
    {
    }

    protected override void OnFrameApply(ImageFrame<Rgba32> source)
    {
        _ = source.TryGetSinglePixelSpan(out Span<Rgba32> span);

        using ReadWriteTexture2D<Rgba32, Vector4> texture =
            Gpu.Default.AllocateReadWriteTexture2D<Rgba32, Vector4>(span, source.Width, source.Height);

        // Process the image on the GPU...

        texture.GetData(span);
    }
}
```

And then use it like so:

```csharp
using Image<Rgba32> image = Image.Load("cat.jpg");

image.Mutate(c => c.ApplyProcessor(new ComputeSharpProcessor()));

image.Save("cat_processed.jpg");
```

This is just not possible at all today, because all the `Clone` and `Mutate` extensions completely lose type information on the image in use (even if the image is explicitly a generic version), so you just can't insert your pixel-specific processor in there.

### Summary

I think this would be a relaativey easy to implement but powerful feature to allow ImageSharp to be more flexible in these more advanced scenarios. I would be happy to help with the actual implementation if we decide to do this, of course 😄

**Comment by JimBobSquarePants** [comment, marked as answer]
>current APIs are effectively causing consumers to lose type information when they're actually working with pixel-specific images.

What type information? Nothing is lost ever. If you want to decode to a specific pixel format, use that format.

```c#
 if (typeof(TPixel) == typeof(Rgba32)) { ... }
 else { ... }
```

**Reply by DaZombieKiller** [reply]
I could be misunderstanding, but I'm assuming this is referring to how you must have a generic `TPixel` (due to how `IImageProcessor` works), instead of the specific pixel type in use by the `Image`. The type information isn't *lost* per se, but instead obscured and requires hacks (eg, `Unsafe.As`) to make it more specific again. Being able to pass in an `IImageProcessor<TPixel>` to `ApplyProcessor` would allow you to avoid that for situations where you only care about a single pixel format. ~~I feel that the methods on `PixelOperations` might be the intended solution for this problem though (they would effectively be a no-op if the pixel formats match, right?).~~

**Reply by DaZombieKiller** [reply]
The problem in the OP could be hacked around without ImageSharp changes by using a type like this:
```CSharp
public abstract class ImageProcessorFactory<TPixel> : IImageProcessor
    where TPixel : unmanaged, IPixel<TPixel>
{
    public abstract IImageProcessor<TPixel> CreateProcessor(Configuration configuration, Image<TPixel> source, Rectangle sourceRectangle);

    IImageProcessor<TPixel2> IImageProcessor.CreatePixelSpecificProcessor<TPixel2>(Configuration configuration, Image<TPixel2> source, Rectangle sourceRectangle)
    {
        if (typeof(TPixel) != typeof(TPixel2))
            throw new InvalidOperationException();

        return Unsafe.As<IImageProcessor<TPixel2>>(CreateProcessor(configuration, Unsafe.As<Image<TPixel>>(source), sourceRectangle));
    }
}
```
Which can then be subclassed as `ComputeSharpProcessor.Factory` and passed to `ApplyProcessor`. However, this lacks the compile-time checks you'd get from a `Mutate(IImageProcessingContext<TPixel>)` overload. As a result, it's not an ideal solution.

**Reply by JimBobSquarePants** [reply]
Just to prove a point....

```c#
public sealed class ComputeSharpProcessor : ImageProcessor<Rgba32>, IImageProcessor
{
    public ComputeSharpProcessor(Image<Rgba32> source)
        : this(source.GetConfiguration(), source, source.Bounds())
    {
        // This constructor only exists to satisfy the base class
        // and ensure that the class usage is type safe.
    }

    private ComputeSharpProcessor(
        Configuration configuration,
        Image<Rgba32> source,
        Rectangle sourceRectangle)
        : base(configuration, source, sourceRectangle)
    {
    }

    public IImageProcessor<TPixel> CreatePixelSpecificProcessor<TPixel>(
        Configuration configuration,
        Image<TPixel> source,
        Rectangle sourceRectangle)
        where TPixel : unmanaged, IPixel<TPixel>
    {
        if (typeof(TPixel) != typeof(Rgba32))
        {
            throw new NotSupportedException();
        }

        // We return a new instance since the source and rectangle are actually
        // the destination following cloning and chaining.
        return (IImageProcessor<TPixel>)
            (object)(new ComputeSharpProcessor(configuration, (Image<Rgba32>)(object)source, sourceRectangle));
    }

    protected override void OnFrameApply(ImageFrame<Rgba32> source)
    {
        throw new NotImplementedException();
    }
}

class Program
{
    static void Main(string[] args)
    {
        Console.WriteLine("Hello World!");

        using var image = new Image<Rgba32>(100, 100);
        image.Mutate(x => x.ApplyProcessor(new ComputeSharpProcessor(image)));
    }
}
```

![image](https://user-images.githubusercontent.com/385879/103609320-88070100-4f15-11eb-9d06-e272dd25609c.png)


This does have compile time checks because the image type parameter has to be `Rgba32`.

Sure it ain't pretty but absolutely not impossible. It also only works for non-cloning processors.

**Reply by JimBobSquarePants** [reply]
I'm not negating the proposal btw, I just want accuracy.

**Reply by JimBobSquarePants** [reply]
`IImageProcessingContext` is already public.

**Reply by JimBobSquarePants** [reply]
Actually. Cloning processors work also.

```c#
public sealed class ComputeSharpCloningProcessor : CloningImageProcessor<Rgba32>, ICloningImageProcessor
{
    public ComputeSharpCloningProcessor(Image<Rgba32> source)
        : this(source.GetConfiguration(), source, source.Bounds())
    {
        // This constructor only exists to satisfy the base class
        // and ensure that the class usage is type safe.
    }

    private ComputeSharpCloningProcessor(
        Configuration configuration,
        Image<Rgba32> source,
        Rectangle sourceRectangle)
        : base(configuration, source, sourceRectangle)
    {
    }

    public ICloningImageProcessor<TPixel> CreatePixelSpecificCloningProcessor<TPixel>(
        Configuration configuration,
        Image<TPixel> source,
        Rectangle sourceRectangle) where TPixel : unmanaged, IPixel<TPixel>
    {
        if (typeof(TPixel) != typeof(Rgba32))
        {
            throw new NotSupportedException();
        }

        // We return a new instance since the source and rectangle are actually
        // the destination following cloning and chaining.
        return (ICloningImageProcessor<TPixel>)
            (object)(new ComputeSharpCloningProcessor(configuration, (Image<Rgba32>)(object)source, sourceRectangle));
    }

    IImageProcessor<TPixel> IImageProcessor.CreatePixelSpecificProcessor<TPixel>(
        Configuration configuration,
        Image<TPixel> source,
        Rectangle sourceRectangle)
        => this.CreatePixelSpecificCloningProcessor(configuration, source, sourceRectangle);

    protected override Size GetDestinationSize()
        => new Size(this.Source.Width / 2, this.Source.Height / 2);

    protected override void OnFrameApply(ImageFrame<Rgba32> source, ImageFrame<Rgba32> destination)
    {
        throw new NotImplementedException();
    }
}

class Program
{
    static void Main(string[] args)
    {
        Console.WriteLine("Hello World!");

        using var image = new Image<Rgba32>(100, 100);
        image.Mutate(x => x.ApplyProcessor(new ComputeSharpCloningProcessor(image)));
    }
}
```

![image](https://user-images.githubusercontent.com/385879/103610112-3d868400-4f17-11eb-93da-ff499019f925.png)

**Reply by DaZombieKiller** [reply]
> 
> 
> `IImageProcessingContext` is already public.

I think that's a typo in the OP (based on the loose `), I believe it's supposed to say:
> Expose `IImageProcessingContext<TPixel>`

**Reply by JimBobSquarePants** [reply]
That should work for both `Clone` and `Mutate`

**Reply by Sergio0694** [reply]
> What type information? Nothing is lost ever.

What I meant is that regardless of if you staart with an `Image` or an `Image<TPixel>`, when you do `Mutate` you always just get the untyped `IImageProcessingContext`, so the type information is "lost" there. You can't add pixel-specific processors in a pipeline.

Regarding your workaround, yeah I do agree that there are ways to make it work already, but that's clunky at best.
You have an empty constructor just to get build-time checks, you allocate a closure because you're capturing the image, you need to have those ugly `(T)(object)` casts to bypass the C# restrictions, you need to have runtime type checks and throw exceptions, you have that weird double implementation of `CreatePixelSpecificProcessor`, etc.
I just feel like having built-in support for this would be cleaner, less error prone and easier to implement for consumers 🤔

**Reply by JimBobSquarePants** [reply]
Oh definitely ugly as sin! I just wanted to prove it wasn’t impossible :smile:

**Comment by antonfirsov** [comment]
@Sergio0694 what's the user value on utilizing the Mutate/Clone infra for this? According to your example, you most likely only want to mutate and never clone. Why don't you just define an extension method over `Image<Rgba32>` within your library?

In cases you want to clone the image you can implement your extension method to return a new `Image<Rgba32>`. I would assume you want strong control over your memory anyways, so it shouldn't be an issue.

**Reply by Sergio0694** [reply]
> what's the user value on utilizing the Mutate/Clone infra for this?

Well, one would be that it'd allow consumers to easily integrate their custom processors into the existing infrastructure.
My idea is that `IImageProcessingContext<TPixel>` would extend `IImageProcessingContext`, so you could do stuff like:

```csharp
using Image<Rgba32> image = Image.Load<Rgba32>("cat.jpg");

image.Mutate(c => c
    .ApplyProcessor(new MyRgba32Processor())
    .Grayscale()
    .Resize(...));
```

As in, you could easily mix custom pixel-specific processors with existing ones in the same pipeline.
That'd be one of the main advantages, and it'd make the whole thing pretty consistent.

Also to be clear, while it's true that this would be nice to have in my library, the proposal was more of a general idea that I actually had for a while about ImageSharp, that I think that this feature would be a nice addition outside of my specific use case in particular 🙂

**Reply by antonfirsov** [reply]
If I understand it correctly, for this kind of integration `ComputeSharpProcessor` will always:
(1) Create a texture (2) copy `Image<Rgba32>` contents to the texture (3) run a shader (4) copy texture back to `Image<Rgba32>`.

Seems rather expensive to me, are you sure it's worth it?

**Reply by Sergio0694** [reply]
Well, it depends on the scenario and the effect of course, but it can still be absolutely worth it, yeah 😄

For instance, consider this:
- 4K image (3840x2560)
- Gaussian blur with radius 80

This takes about **13.5s** to run using ImageSharp's `GaussianBlur` processor, and... **0.6s** using ComputeSharp.
That's **225x** faster, even including the time to copy the texture to GPU, dispatch the shader and copy it back 🚀
And it could also optimized further on my end as well, this was really just a very simple demonstration I use to illustrate my lib 😊

> **NOTE:** I have a Ryzen 7 2700x, 16GB of DDR4 @ 3200Mhz C16 RAM, and a GTX1080

**Reply by JimBobSquarePants** [reply]
Is that 13.5s with all our changes? I know GPUs are designed for this kind of thing but that makes me sad.

**Reply by JimBobSquarePants** [reply]
@Sergio0694   Just to clarify, in your original example. How is this constructor called?

```c#
image.Mutate(c => c.ApplyProcessor(new ComputeSharpProcessor()));
```
`ImageProcessor<TPixel>` has an explicit constructor requiring parameters.

Additionally, how do you chain with existing `IImageProcessingContext` extensions without multiple `Mutate|Clone` calls

- Expose `Clone` and `Mutate` extensions taking `Action<IImageProcessingContext<TPixel>>`
- Expose `Clone` and `Mutate` extensions taking `params IImageProcessor<TPixel>[]`

Those methods restrict `Mutate|Clone` to the more constrained interface.

**Reply by Sergio0694** [reply]
Alright so, I got tricked by the `GaussianBlur` extension, which only exposes the `float sigma` parameter, not the `int radius` one 🤦‍♂️
So that 13s time was basically a combination of me using the previous ImageSharp build + that extension resulting in a much bigger convolution radius. I'm not using the processor directly, which has that overload (related, do we want to expose that through an extension too?). Here are the updated results, also with the latest ImageSharp version. It fares much better in comparison now 😄

| Method | Categories |        Mean |     Error |    StdDev | Ratio |
|--------|----------- |------------ |----------:|----------:|------:|
|    Cpu |      BOKEH | 2,734.88 ms | 35.000 ms | 23.151 ms |  1.00 |
|    Gpu |      BOKEH |   169.80 ms |  7.838 ms |  5.185 ms |  0.06 |
|        |            |             |           |           |       |
|    Cpu |   GAUSSIAN |   523.10 ms |  7.018 ms |  4.642 ms |  1.00 |
|    Gpu |   GAUSSIAN |    62.09 ms |  2.647 ms |  1.751 ms |  0.12 |

The GPU version is still much faster of course, and also remember that it's not really that optimized either.
That said, ImageSharp doesn't look too terrible in comparison, especially with the gaussian blur.
Back to Anton's question though, yeah in these cases for instance it'd still be worth it to copy and process on the GPU.
Though as I mentioned, the idea to expose these pixel-specific APIs was a general proposal, not strictly related to my library 😊

**Reply by JimBobSquarePants** [reply]
OK. So there's a LOT more to this than initially meets the eye. 

In order to make the method typesafe you require a generic constraint of some form for the processor. There's a two ways to do this:

 1. Add an `IImageProcessingContext<TPixel>` type with explicit `Mutate|Clone` methods as described by @Sergio0694 .
 2. Pass something containing `TPixel` as a type parameter to the processor constructor. My hack above.
 
 There are issues with both approaches:

The first potentially works (_though I still don't know where the missing parameterless constructor comes from_) disallows mixing pixel type specific with non specific processors due to the new context constraints. I.E this code would not compile.

```c#
using var image = new Image<Rgba32>(100, 100);
image.Mutate(
    x => x.Crop(new Rectangle(1, 1, 99, 99)).ApplyProcessor(new Rgba32ImageProcessor()));
```

The second provably works and allows mixing but requires some funkiness (_to put it mildly_) in the processor itself by creating the requirement for two constructors.

This is the tightest version I could come up with (_Consumer would write this_):

```c#
public sealed class Rgba32ImageProcessor : PixelTypeSpecificImageProcessor<Rgba32, Rgba32ImageProcessor>
{
    public Rgba32ImageProcessor(Image<Rgba32> source)
        : base(source)
    {
    }

    private Rgba32ImageProcessor(Configuration configuration, Image<Rgba32> source, Rectangle sourceRectangle)
        : base(configuration, source, sourceRectangle)
    {
    }

    protected override Rgba32ImageProcessor CreatePixelSpecificProcessor(
        Configuration configuration,
        Image<Rgba32> source,
        Rectangle sourceRectangle)
        => new Rgba32ImageProcessor(configuration, source, sourceRectangle);

    protected override void OnFrameApply(ImageFrame<Rgba32> source)
    {
        throw new NotImplementedException();
    }
}
```

Where the base class looks like this (_Would be implemented in ImageSharp_):

```c#
public abstract class PixelTypeSpecificImageProcessor<TPixel, TProcessor> : ImageProcessor<TPixel>, IImageProcessor
    where TPixel : unmanaged, IPixel<TPixel>
    where TProcessor : ImageProcessor<TPixel>
{
    protected PixelTypeSpecificImageProcessor(Image<TPixel> source)
        : this(source.GetConfiguration(), source, source.Bounds())
    {
        // This constructor only exists to satisfy the base class
        // and ensure that the class usage is type safe.
        // The assigned values are not used during processing.
    }

    protected PixelTypeSpecificImageProcessor(
        Configuration configuration,
        Image<TPixel> source,
        Rectangle sourceRectangle)
        : base(configuration, source, sourceRectangle)
    {
    }

    protected abstract TProcessor CreatePixelSpecificProcessor(
        Configuration configuration,
        Image<TPixel> source,
        Rectangle sourceRectangle);

    IImageProcessor<TPixel1> IImageProcessor.CreatePixelSpecificProcessor<TPixel1>(
        Configuration configuration,
        Image<TPixel1> source,
        Rectangle sourceRectangle)
    {
        // We add this check to ensure no-one tries to create a processor
        // outwith the normal pipeline. The types should always match due to compiler
        // restrictions in normal environments.
        if (typeof(TPixel1) != typeof(TPixel))
        {
            throw new NotSupportedException($"Type parameter must be {typeof(TPixel)}.");
        }

        // We return a new instance since the source and rectangle are actually
        // the destination following cloning and chaining.
        return (IImageProcessor<TPixel1>)CreatePixelSpecificProcessor(configuration, (Image<TPixel>)(object)source, sourceRectangle);
    }
}
```

Usage would be as follows:

```c#
using var image = new Image<Rgba32>(100, 100);
image.Mutate(
    x => x.Crop(new Rectangle(1, 1, 99, 99)).ApplyProcessor(new Rgba32ImageProcessor(image)));
```
 
It's certainly odd but I cannot think of a better way to implement the approach and am open to ideas.

**Reply by antonfirsov** [reply]
I have an alternative idea, will elaborate tomorrow.

**Reply by Sergio0694** [reply]
Very curious to hear Anton's take on this 😄

My general idea was (just as a very rough draft) to do something like this, first add an interface like:

```csharp
public interface IImageProcessingContext<TPixel> : IImageProcessingContext
    where TPixel : unmanaged, IPixel<TPixel>
{
    new IImageProcessingContext<TPixel> ApplyProcessor(IImageProcessor processor, Rectangle rectangle);
    new IImageProcessingContext<TPixel> ApplyProcessor(IImageProcessor processor);
    IImageProcessingContext<TPixel> ApplyProcessor(IImageProcessor<TPixel> processor, Rectangle rectangle);
    IImageProcessingContext<TPixel> ApplyProcessor(IImageProcessor<TPixel> processor);
}
```

Then we could add new overloads to the various extensions, like this for `Crop` (including existing extensions):

```csharp
public static class CropExtensions
{
    public static IImageProcessingContext Crop(this IImageProcessingContext source, int width, int height);
    public static IImageProcessingContext Crop(this IImageProcessingContext source, Rectangle cropRectangle);
    public static IImageProcessingContext<TPixel> Crop<TPixel>(this IImageProcessingContext<TPixel> source, int width, int height)
        where TPixel : unmanaged, IPixel<TPixel>;
    public static IImageProcessingContext<TPixel> Crop<TPixel>(this IImageProcessingContext<TPixel> source, Rectangle cropRectangle)
        where TPixel : unmanaged, IPixel<TPixel>;
}
```

So with this, you'd be able to retain type info when chaining processors, including combining pixel-specific ones and not.
Then when you wanted to insert a pixel-specific one into a `Mutate` chain, you could just use `ApplyProcessor`, picking the overload taking `IImageProcessor<TPixel>`, and that'd basicalaly be similar to the non pixel-specific one, in that it'd take a first instance that just acts as a definition for the processor (but is still strongly typed to the pixel type), and then from there it'd invoke an interface method to actually get the single-use instance with the target image. Basically you'd still keep the pattern of the two processor instances we have today (one for the definition, one for the application). This would allow us to avoid clunky solutions like that empty constructor that captured the target image in a lambda expression, for instance.

As I said, just a general thought, will wait to see what Anton's proposal is first before elaborating on this further 😄
Really really happy to see more interest in this feature though! 🎉

**Reply by JimBobSquarePants** [reply]
Yeah, let's see what @antonfirsov comes up with. That's a LOT of extensions to update across both here and Drawing!!

**Reply by antonfirsov** [reply]
Well, my "idea" is quite dumb actually. I still don't think introducing pixel-specific processors is the right way to solve this.
I believe, that in this area our first priority should be to maintain simplicity and clarity, and supporting advanced scenarios only comes as second. Currently Image Processor is an abstraction for an operation that can be executed on any image regardless of the pixel type. I think it would serve our users the best if we keep this concept as-is.

My point is that - even with the above constraints - **there is still space** to support shader-based processors (actually in a better way IMO) by implementing multiple kernels for different pixel types and and having automatic conversion for the unknown ones :

```C#
public sealed class ComputeSharpProcessor<TPixel> : ImageProcessor<TPixel>
{
    public ComputeSharpProcessor(Configuration configuration, Image<TPixel> source, Rectangle sourceRectangle)
        : base(configuration, source, sourceRectangle)
    {
    }

    protected override void OnFrameApply(ImageFrame<TPixel> source)
    {
        _ = specific.TryGetSinglePixelSpan(out Span<TPixel> genericSpan);

        if (typeof(TPixel) == typeof(Rgba32))
        {
            Span<Rgba32> span = MemoryMarshal.Cast<TPixel, Rgba32>(genericSpan);
            Texture<Rgba32> texture = Gpu.CreateTexture(span, source.Width, source.Height);
            Kernels.MyKernelRgba32(texture);
            texture.GetData(span);
        }
        else if (typeof(TPixel) == typeof(Rgb24))
        {
            Span<Rgb24> span = MemoryMarshal.Cast<TPixel, Rgb24>(genericSpan);
            Texture<Rgb24> texture = Gpu.CreateTexture(span, source.Width, source.Height);
            
            // This kernel could be a slight alteration of "MyKernelRgba32" that starts by concerting the input pixel, and reuses the core logic
            Kernels.MyKernelRgb24(texture);
            texture.GetData(span);
        }
        // else if ...
        else
        {
            // Ideally, this won't be called often, but note that most conversions should be very fast since #1404

            using var tempBuffer = this.Configuration.MemoryAllocator.Allocate<Rgba32>(genericSpan.Length);
            Span<Rgba32> span = tempBuffer.Span;
            PixelOperations<TPixel>.ToRgba32(this.Configuration, genericSpan, span);

            Texture<Rgba32> texture = Gpu.CreateTexture(span, source.Width, source.Height);
            Kernels.MyKernelRgba32(texture);
            texture.GetData(span);

            PixelOperations<TPixel>.ToRgba32(this.Configuration, span, genericSpan);
        }
    }
}
```

**Reply by antonfirsov** [reply]
> Then we could add new overloads to the various extensions, like this for Crop (including existing extensions) [...]

I really don't think we should ever do this. This pollutes API for users a lot, and creates maintainability hell for us.

**Reply by JimBobSquarePants** [reply]
This is the way. 

Yep totally agree. This would have been my first choice approach and I only put together the class I demonstrated to prove a point.

**Reply by Sergio0694** [reply]
Right, yeah those are all very good points, I have to agree that doing this would be pretty complex 😄

I have to say I'm still not a fan of hiding the type information though - while I do understand that having a non-generic interface for pixel-format agnostic processor is super nice, I don't particularly like forcing this on consumers. Admittedly though, I might be a bit biased on this because lately I've been working quite a bit with Roslyn APIs in my project, and those are really all-in with generics everywhere, and they have this aspect of "not losing type information" all over the place, as tons and tons of methods are generic specifically to be able to return the same type as the input (when used in a fluent chain call like in ImageSharp).

As mentioned though, yeah I can see how that'd require a big number of new overloads for a feature that'd probably be used just by a small subset of users, that's fair. Still very glad we could have this conversation though! 😊

![](https://user-images.githubusercontent.com/10199417/104063221-20460400-51fc-11eb-8cd7-b37aa6eb6e69.gif)
